pills/09-automatic-runtime-dependencies.md

```diff
@@ -46,14 +46,16 @@ We now note that Nix automatically recognized build dependencies once our `deriv
 
 Nix handles runtime dependencies for us automatically. The technique it uses to do so may seem fragile at first glance, but it works so well that the NixOS operating system is built off of it. The underlying mechanism relies on the hash of the store paths. It proceeds in three steps:
 
-1.  Dump the derivation as a NAR. Recall that this is a serialization of the derivation output \-- meaning this works fine whether the output is a single file or a directory.
+1.  Dump the derivation as a NAR. Recall that this is a serialization of the derivation output — meaning this works fine whether the output is a single file or a directory.
 
-2.  For each build dependency, search the contents of the NAR for the hash part of its out path. For example:
+2.  For each build dependency (including transitive dependencies), search the contents of the NAR for the hash part of its out path. For example:
     * Our derivation depends on `/nix/store/sk590g7fv53m3zp0ycnxsc41snc2kdhp-gzip-1.6.drv`
     * Its output path is `/nix/store/hcrf95x3r60kw71wgwbdybjfcq0ipkpj-gzip-1.6`.
     * Therefore, Nix will search the NAR for `hcrf95x3r60kw71wgwbdybjfcq0ipkpj`.
 
-3.  If the hash is found, then it's a runtime dependency.
+3.  Each hash which is found somewhere in the NAR is recorded as a runtime dependency.
+
+(For completeness: some derivations have multiple output paths. In that case, Nix will search for the hashes of all the outputs. Also, Nix will search for the hashes of source dependencies, such as our `build.sh` file. The authoritative definition is the [source code](https://github.com/NixOS/nix/blob/a268c0de7192188c7233bf83a4635198c360e270/src/libstore/build/local-derivation-goal.cc#L2220-L2227).)
```

Comment:
> hashes of all the outputs

I would expect only the referenced outputs to be looked up but have not verified.

Reply:
Indeed, I just checked (by creating files with hashes in a derivation output). Nix searches for all referenced outputs, plus the derivation's "base output" (I'm not sure what's its official name). It doesn't search for non-referenced outputs. I modified "all the outputs" into "all the referenced outputs".

(How I tested this:

I created a derivation with multiple outputs by adding `outputs = [ "bin" "dev" "out" ];` into `hello.nix`, and then I get:

```
❯ nix-store -r /nix/store/p6cmf63lgjfmx6vn0zsk790x8b37ym1l-hello.drv
/nix/store/3x15z8r1gdrmdq3zlnhvrfnymw2jcq43-hello
/nix/store/hhx3wqmla53xyh88nx3il7cq3bbqdy1y-hello-bin
/nix/store/q453yisz0g5dhrjbadmfxf0m3m8ncpc6-hello-dev
```

I added the plain `hello` derivation as an input to the `simple` derivation, and `nix derivation show` shows that it depends on the `bin` output. I see that if I write the hash of the `-bin` output it's added as a dependency, and if I add the hash of the no-suffix output it's added as a dependency, but if I add the hash of the `-dev` output it's not added as a dependency.
)